--- etc/freshclam.conf.orig	2007-02-14 12:26:13.481124568 -0600
+++ etc/freshclam.conf	2007-02-14 12:26:55.021809424 -0600
@@ -5,7 +5,7 @@
 
 
 # Comment or remove the line below.
-Example
+#Example
 
 # Path to the database directory.
 # WARNING: It must match clamd.conf's directive!
@@ -14,7 +14,7 @@
 
 # Path to the log file (make sure it has proper permissions)
 # Default: disabled
-#UpdateLogFile /var/log/freshclam.log
+UpdateLogFile /var/log/freshclam.log
 
 # Enable verbose logging.
 # Default: no
