--- ../clamav-0.90-orig/clamd/clamd.c	2007-02-12 14:55:24.000000000 -0600
+++ ./clamd/clamd.c	2007-02-13 22:04:32.000000000 -0600
@@ -229,11 +229,16 @@
     if((cpt = cfgopt(copt, "LogFile"))->enabled) {
 	logg_file = cpt->strarg;
 	if(strlen(logg_file) < 2 || (logg_file[0] != '/' && logg_file[0] != '\\' && logg_file[1] != ':')) {
+	if (strcmp(logg_file,"stderr")!=0) {
 	    fprintf(stderr, "ERROR: LogFile requires full path.\n");
 	    logg_close();
 	    freecfg(copt);
 	    return 1;
 	}
+	else {
+		use_stderr=1;
+	}
+	}
 	time(&currtime);
 	if(logg("#+++ Started at %s", ctime(&currtime))) {
 	    fprintf(stderr, "ERROR: Problem with internal logger. Please check the permissions on the %s file.\n", logg_file);
--- ../clamav-0.90-orig/shared/output.c	2006-08-30 17:41:21.000000000 -0500
+++ ./shared/output.c	2007-02-13 22:03:48.000000000 -0600
@@ -138,7 +138,9 @@
     pthread_mutex_lock(&logg_mutex);
 #endif
     if(logg_file) {
-	if(!logg_fd) {
+	if(use_stderr) {
+	   logg_fd = stderr;
+	} else if(!logg_fd) {
 	    old_umask = umask(0037);
 	    if((logg_fd = fopen(logg_file, "a")) == NULL) {
 		umask(old_umask);
@@ -161,7 +163,7 @@
 	    }
 	}
 
-	if(logg_size) {
+	if(logg_size && !use_stderr) {
 	    if(stat(logg_file, &sb) != -1) {
 		if(sb.st_size > logg_size) {
 		    logg_file = NULL;
--- ../clamav-0.90-orig/shared/output.h	2007-02-13 15:14:38.000000000 -0600
+++ ./shared/output.h	2007-02-13 22:04:14.000000000 -0600
@@ -29,6 +29,7 @@
 
 int mdprintf(int desc, const char *str, ...);
 
+int use_stderr;
 int logg(const char *str, ...);
 
 void logg_close(void);
