diff -u ./ChangeLog ../qmailmrtg7-3.6-shupp/ChangeLog
--- ./ChangeLog	Sat Apr  6 08:37:09 2002
+++ ../qmailmrtg7-3.6-shupp/ChangeLog	Mon Aug 25 17:16:25 2003
@@ -1,3 +1,9 @@
+3.6 - shupp
+08/25/03 - change msg graphs back to msg, as well as pop/smtp totals
+         - added d, T, S for dnscache, tinydns, and spamd reporting 
+           respectively
+         - removed duplicate usage statement
+
 3.6
 03/11/02 - change msg graphs to msg/hour
          - set concurrency lable on graphs to max concurrency
Only in .: checkq
diff -u ./qmail.mrtg.cfg ../qmailmrtg7-3.6-shupp/qmail.mrtg.cfg
--- ./qmail.mrtg.cfg	Sat Apr  6 08:37:08 2002
+++ ../qmailmrtg7-3.6-shupp/qmail.mrtg.cfg	Mon Aug 25 17:15:32 2003
@@ -9,7 +9,7 @@
 Target[msg]: `/usr/local/bin/qmailmrtg7 m /var/log/qmail`
 PageTop[msg]: <B>FQDN Messages</B><br>
 ShortLegend[msg]: msg
-YLegend[msg]: msg/hour
+YLegend[msg]: msg
 Legend1[msg]: Total Msg&nbsp;
 LegendI[msg]: Deliveries:&nbsp;
 LegendO[msg]: Attempts:&nbsp;
@@ -54,7 +54,7 @@
 Target[messstatus]: `/usr/local/bin/qmailmrtg7 s /var/log/qmail`
 PageTop[messstatus]: <B>FQDN - Message Status</B><BR> 
 ShortLegend[messstatus]: msg
-YLegend[messstatus]: msg/hour
+YLegend[messstatus]: msg
 Legend1[messstatus]: msg&nbsp;
 LegendI[messstatus]: success&nbsp;
 LegendO[messstatus]: failures&nbsp;
@@ -99,7 +99,7 @@
 Target[smtpad]: `/usr/local/bin/qmailmrtg7 a /var/log/smtp`
 PageTop[smtpad]: <B>FQDN - SMTP Totals</B><BR>
 ShortLegend[smtpad]: SMTP
-YLegend[smtpad]: SMTP/hour
+YLegend[smtpad]: SMTP
 Legend1[smtpad]: SMTP&nbsp;
 LegendI[smtpad]: Allow&nbsp;
 LegendO[smtpad]: Deny&nbsp;
@@ -128,9 +128,61 @@
 Target[pop3ad]: `/usr/local/bin/qmailmrtg7 a /var/log/pop3`
 PageTop[pop3ad]: <B>FQDN - POP3 Totals</B><BR>
 ShortLegend[pop3ad]: POP3
-YLegend[pop3ad]: POP3/hour
+YLegend[pop3ad]: POP3
 Legend1[pop3ad]: POP3&nbsp;
 LegendI[pop3ad]: Allow&nbsp;
 LegendO[pop3ad]: Deny&nbsp;
 WithPeak[pop3ad]: ymwd
 
+#
+# the target below is for spamd (via multilog)
+#
+##-------------------------------------------------------------------
+#
+#Title[spam]: Spam - FQDN
+#MaxBytes[spam]: 120
+#AbsMax[spam]: 120
+#Options[spam]: gauge
+#Target[spam]: `/usr/local/bin/qmailmrtg7 S /service/spamd/log/main`
+#PageTop[spam]: <B>FQDN Spam</B><br>
+#ShortLegend[spam]: 5 minute totals
+#YLegend[spam]: spam scanning
+#Legend1[spam]: Total Messages&nbsp;
+#LegendI[spam]: spam&nbsp;identified:&nbsp;
+#LegendO[spam]: clean&nbsp;message:&nbsp;
+#WithPeak[spam]: ymwd
+
+#
+# the 2 targets below are for djbdns
+#
+##-------------------------------------------------------------------
+#
+#Title[dnscache]: FQDN - dnscache Totals
+#MaxBytes[dnscache]: 1000
+#AbsMax[dnscache]: 500000
+#Options[dnscache]: gauge
+#Target[dnscache]: `/usr/local/bin/qmailmrtg7 d /service/dnscachex/log/main/`
+#PageTop[dnscache]: <B>FQDN - dnscache Totals</B><BR>
+#ShortLegend[dnscache]: dns queries/5 min
+#YLegend[dnscache]: dns queries
+#Legend1[dnscache]: Total Queries
+#LegendI[dnscache]: &nbsp;total queries:
+#LegendO[dnscache]: &nbsp;failures:
+#WithPeak[dnscache]: ymwd
+#
+##-------------------------------------------------------------------
+#
+#Title[tinydns]: FQDN - tinydns Responses/Drops
+#MaxBytes[tinydns]: 1000
+#AbsMax[tinydns]: 500000
+#Options[tinydns]: gauge,growright
+#Target[tinydns]: `/usr/local/bin/qmailmrtg7 d /var/log/tinydns/`
+#PageTop[tinydns]: <B>FQDN - tinydns Totals</B><BR>
+#ShortLegend[tinydns]: queries/5 min
+#YLegend[tinydns]: dns queries
+#Legend1[tinydns]: Successful Queries
+#Legend2[tinydns]: Non-Authoritative Drops
+#LegendI[tinydns]: &nbsp;responses:
+#LegendO[tinydns]: &nbsp;drops:
+#WithPeak[tinydns]: ymwd
+#
Only in .: qmailmrtg7
diff -u ./qmailmrtg7.c ../qmailmrtg7-3.6-shupp/qmailmrtg7.c
--- ./qmailmrtg7.c	Sat Apr  6 08:37:08 2002
+++ ../qmailmrtg7-3.6-shupp/qmailmrtg7.c	Mon Aug 25 17:07:34 2003
@@ -41,6 +41,12 @@
 int local;
 int remote;
 int max_files;
+int dnscachequery;
+int nxdomain;
+int tinydnsresponse;
+int tinydnsdrop;
+int tspam;
+int tnotspam;
 
 time_t end_time;
 time_t start_time;
@@ -62,10 +68,6 @@
 
     if ( argc != 3 ) { 
         usage();
-        printf("usage: type dir\n");
-        printf("where type is one of t, a, m, c, s, b, q, r, R \n");
-        printf("and dir is a directory containing multilog files\n");
-        printf("for q option dir is the qmail queue dir\n");
         exit(-1);
     }
     strncpy(TheDir, argv[2], MAX_BUFF);
@@ -81,6 +83,9 @@
       case 'q':
       case 'r':
       case 'R':
+      case 'd':
+      case 'S':
+      case 'T':
         break;
       default:
         usage();
@@ -152,12 +157,27 @@
 
         /* tcpserver allow/deny */
         case 'a':
-            printf("%d\n%d\n\n\n", tallow*12, tdeny*12);
+            printf("%d\n%d\n\n\n", tallow, tdeny);
             break;
 
         /* messages */
         case 'm':
-            printf("%d\n%d\n\n\n",success*12, (failure+success)*12);
+            printf("%d\n%d\n\n\n",success, (failure+success));
+            break;
+
+        /* dnscache queries */
+        case 'd':
+            printf("%d\n%d\n\n\n", dnscachequery, nxdomain);
+            break;
+
+        /* tinydns queries */
+        case 'T':
+            printf("%d\n%d\n\n\n", tinydnsresponse, tinydnsdrop);
+            break;
+
+        /* spamd */
+        case 'S':
+            printf("%d\n%d\n\n\n", tspam, tnotspam);
             break;
 
         /* remote/local concurrency */
@@ -235,6 +255,33 @@
             }
             break;
 
+        case 'd':
+            if ((tmpstr1 = strstr(TmpBuf, " query "))!=NULL) {
+                ++dnscachequery;
+            }
+            if ((tmpstr1 = strstr(TmpBuf, " nxdomain "))!=NULL) {
+                ++nxdomain;
+            }
+            break;
+
+        case 'T':
+            if ((tmpstr1 = strstr(TmpBuf, " + "))!=NULL) {
+                ++tinydnsresponse;
+            }
+            if ((tmpstr1 = strstr(TmpBuf, " - "))!=NULL) {
+                ++tinydnsdrop;
+            }
+            break;
+
+        case 'S':
+            if ((tmpstr1 = strstr(TmpBuf, " clean message "))!=NULL) {
+                ++tnotspam;
+            }
+            if ((tmpstr1 = strstr(TmpBuf, " identified spam "))!=NULL) {
+                ++tspam;
+            }
+            break;
+
         case 'c':
             if ((tmpstr1 = strstr(TmpBuf, " status: ")) != NULL ) {
                 while(*tmpstr1!='/'&&*tmpstr1!=0) ++tmpstr1;
@@ -305,7 +352,7 @@
 void usage()
 {
     printf("usage: type dir\n");
-    printf("where type is one of t, a, m, c, s, b, q \n");
+    printf("where type is one of t, a, m, c, s, b, q, d, T, S \n");
     printf("and dir is a directory containing multilog files\n");
     printf("for q option dir is the qmail queue dir\n");
 }
diff -u ./qmailmrtg7.h ../qmailmrtg7-3.6-shupp/qmailmrtg7.h
--- ./qmailmrtg7.h	Sat Apr  6 08:37:09 2002
+++ ../qmailmrtg7-3.6-shupp/qmailmrtg7.h	Mon Aug 25 17:07:37 2003
@@ -1,2 +1,2 @@
-int BigTodo=1;
-int ConfSplit=53;
+int BigTodo=0;
+int ConfSplit=23;
