--- skel/install/anomalies.orig	Tue Aug  8 11:01:13 2006
+++ skel/install/anomalies	Tue Aug  8 11:01:52 2006
@@ -30,4 +30,5 @@
 # PARENT_RE is a regular expression used to verify that we have a valid parent
 # directory.  If invalid, try ascending again until we run out of directories
 # or match.
-PARENT_RE = "^" + VPop + "/domains/[^/]+\.[^/]+$"
+#PARENT_RE = "^" + VPop + "/domains/[^/]+\.[^/]+$"
+PARENT_RE = "^" + VPop + "/domains/[\w\/]*[^/]+\.[^/]+$"
--- skel/uninstall/anomalies.orig	Tue Aug  8 11:05:52 2006
+++ skel/uninstall/anomalies	Tue Aug  8 11:05:58 2006
@@ -18,4 +18,5 @@
 # PARENT_RE is a regular expression used to verify that we have a valid parent
 # directory.  If invalid, try ascending again until we run out of directories
 # or match.
-PARENT_RE = "^" + VPop + "/domains/[^/]+\.[^/]+$"
+#PARENT_RE = "^" + VPop + "/domains/[^/]+\.[^/]+$"
+PARENT_RE = "^" + VPop + "/domains/[\w\/]*[^/]+\.[^/]+$"
